doc/man7/openssl-core.h.pod

```diff
@@ -0,0 +1,251 @@
+
+The parameter data is an integer (signed or unsigned) of arbitrary
+length, organized in native form, i.e. most significant byte first on
+Big-Endian systems, and least significant byte first on Little-Endian
+systems.
+
+=item C<OSSL_PARAM_REAL>
+
+=for comment It's still debated if we need this or not.
+
+The parameter data is a floating point value in native form.
+
+=item C<OSSL_PARAM_UTF8_STRING>
+
+The parameter data is a printable string.
+
+The pointer form
+
+=item C<OSSL_PARAM_OCTET_STRING>
+
+The parameter data is an arbitrary string of bytes.
+
+=back
+
+Additionally, this flag can be added to any type:
+
+=over 4
+
+=item C<OSSL_PARAM_POINTER_FLAG>
+
+This can be used to indicate that constant data is or will be passed.
+It works by assuming that the memory pointed at by the C<buffer> field
+contains a pointer to the actual data.
+
+If this C<OSSL_PARAM> is used to set a parameter, C<buffer_size> must
+be set to the size of the data pointed at rather than the size of the
+memory pointed at by the C<buffer> field.
+If this C<OSSL_PARAM> is used in a parameter request, C<buffer_size>
+is not relevant.
+However, the responding function will set C<*return_size> to the size
+of the actual data.
+
+Note that the use of ths flag is B<fragile> and can only be safely
+used for data that remains constant and in a constant location for a
+long enough duration (such as the life-time of the object that
+provides these parameters).
+
+=back
+
+For convenience, these types are provided:
+
+=over 4
+
+=item C<OSSL_PARAM_UTF8_STRING_PTR>
+
+=item C<OSSL_PARAM_OCTET_STRING_PTR>
+
+These are combinations of C<OSSL_PARAM_UTF8_STRING> as well as
+C<OSSL_PARAM_OCTET_STRING> with C<OSSL_PARAM_POINTER_FLAG>.
+
+=back
+
+=back
+
+=back
+
+=head1 EXAMPLES
+
+=head2 OSSL_PARAM
+
+A couple of examples to just show how C<OSSL_PARAM> arrays could be
+set up.
+
+=head3 Example 1
+
+This example is for setting parameters on something:
+
+    #include <openssl/core.h>
+
+    const char *foo = "some string";
+    size_t foo_l = strlen(foo) + 1;
+    const char bar[] = "some other string";
+    const OSSL_PARAM set[] = {
+        { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, foo_l, NULL },
+        { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), NULL },
+        { NULL, 0, NULL, 0, NULL }
+    };
+
+=head3 Example 2
+
+This example is for requesting parameters on something:
+
+    const char *foo = NULL;
+    size_t foo_l;
+    char bar[1024];
+    size_t bar_l;
+    const OSSL_PARAM request[] = {
+        { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, 0 /*irrelevant*/, &foo_l },
+        { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), &bar_l },
+        { NULL, 0, NULL, 0, NULL }
+    };
+
+A responding function that receives this array (as C<params> in this
+example) could fill in the parameters like this:
+
+    /* const OSSL_PARAM *params */
+
+    *(char **)params[0].buffer = "foo value";
+    *params[0].return_size = 10; /* size of "foo value" incl NUL */
```

Comment:
I'd much prefer to see these indexed using the key rather than an integer:
```
ossl_param_set_string(params, "foo", "foo value", 10);
ossl_param_set_string(params, "bar", "var value", 10);
```
note the similarity despite the underlying type being different.

Reply:
The integer is just the position in the array. It's known, why do want to make that more complex? If you use the key, it means that `ossl_param_set_string` will have to search the array. What does that gain us?

Reply:
If we are willing to **guarantee** that the position of an item will **never** change in any param list, integer indices are fine.  I'm not comfortable making that assumption.

Let's future proof this now so we don't create new legacies.

Reply:
Avoiding needing to search can be added if/when necessary.  Worst case should be reducible to one pass over the param array building an index.

Reply:
Ah, I see, I'm taking way too many things for granted.  The example is probably better spelled out like this:

``` C
/* const OSSL_PARAM *params */

int i;

for (i = 0; params[i].key != NULL; i++) {
    if (strcmp(params[i].key, "foo") == 0) {
        *(char **)params[i].buffer = "foo value";
        *params[i].return_size = 10; /* size of "foo value" incl NUL */
    } else if (strcmp(params[i].key, "bar") == 0) {
        memcpy(params[1].buffer, "bar value", 10);
        *params[1].return_size = 10; /* size of "bar value" incl NUL */
    }
    /* Ignore stuff we don't know */
}
```

Reply:
So yeah, it would certainly be possible to create a set of utility macros that helps writing that in a nicer way.

Reply:
That's a better example.  I still think wrapped lookups are nicer for the user, especially in the presence of _PTR and not.  Without any optimisation, there is no complexity difference since both are effectively naïve searches but even there having wrappers can add some safety and nicety.  e.g. "foo" changes from an unsigned integer to an integer when a new flavour of an (currently) unknown algorithm gets introduced.

I had meant to propose something more substantial this week but didn't get to it for a number of reasons.  I will get something done given a bit more time.